File: modules/protocol/base/src/main/java/candle/protocol/codec/AbstractDataCodec.java
package candle.protocol.codec;

import candle.types.resource.Identifier;
import candle.types.resource.Registry;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public abstract class AbstractDataCodec extends Registry<TypeCodec<?>> {

  public AbstractDataCodec( Identifier registryName ) {
    super(registryName);
  }

  protected <T> void writeType( OutputStream out, TypeIdentifier<T> identifier, T value ) throws
                                                                                          IOException {
    @SuppressWarnings("unchecked")
    TypeCodec<T> codec = (TypeCodec<T>) get(identifier);
    if ( codec == null ) {
      throw new IOException("No codec registered with identifier: " + identifier);
    }
    codec.write(out, value, this);
  }

  protected <T> T readType( InputStream in, TypeIdentifier<T> identifier ) throws
                                                                           IOException {
    @SuppressWarnings("unchecked")
    TypeCodec<T> codec = (TypeCodec<T>) get(identifier);
    if ( codec == null ) {
      throw new IOException("No codec registered with identifier: " + identifier);
    }
    return codec.read(in, this);
  }
}

File: protocol/base/src/main/kotlin/net/candlemc/protocol/codec/AbstractDataCodec.kt
package net.candlemc.protocol.codec

import net.candlemc.types.resource.Identifier
import net.candlemc.types.resource.Registry
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

abstract class AbstractDataCodec(registryName: Identifier) : Registry<TypeCodec<*>?>(registryName) {
    @Throws(IOException::class)
    fun <T> writeType(out: OutputStream?, identifier: TypeIdentifier<T>, value: T) {
        val codec: TypeCodec<T> = get(identifier) as TypeCodec<T>?
            ?: throw IOException("No codec registered with identifier: $identifier")
        codec.write(out, value, this)
    }

    @Throws(IOException::class)
    fun <T> readType(`in`: InputStream?, identifier: TypeIdentifier<T>): T {
        val codec: TypeCodec<T> = get(identifier) as TypeCodec<T>?
            ?: throw IOException("No codec registered with identifier: $identifier")
        return codec.read(`in`, this)
    }
}
